dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/XPathMapper.java

```diff
@@ -70,13 +71,101 @@
+   */
+  public static <T> T readDocument(Document doc, Class<T> bindMapperClass, String... rootPaths) {
+    try {
+      XPathRoot rootAnnotation = bindMapperClass.getAnnotation(XPathRoot.class);
+      T bindMapper = bindMapperClass.getDeclaredConstructor().newInstance();
+
+      ensureValidRootPaths(rootAnnotation, rootPaths);
+
+      XPathMapper xPathMapper = new XPathMapper(doc, determineRootPaths(rootAnnotation, rootPaths),
+          determineDefaultRootNamespace(rootAnnotation));
+
+      List<Method> methodsWithAnnotation = ReflectionUtils.getMethodsAnnotatedWith(bindMapperClass, XPathBinding.class);
+      for (Method m : methodsWithAnnotation) {
+        ReflectionUtils.ensureAccess(m);
+        m.invoke(bindMapper, xPathMapper.invoke(bindMapper, m, null));
+      }
+
+      Set<Field> fieldsWithMapper = ReflectionUtils.findFields(bindMapperClass, XPathRoot.class);
+      for (Field f: fieldsWithMapper) {
+        XPathRoot rootBinding = f.getAnnotation(XPathRoot.class);
+
+        // For the embedded mappers, we have to contactenate the root paths
+        String[] subMapperRootPaths = determineRootPaths(rootAnnotation, rootPaths);
+        if (rootBinding != null && rootBinding.value() != null && rootBinding.value().length == 1) {
+          for (int i=0; i<subMapperRootPaths.length; i++) {
+            subMapperRootPaths[i] = subMapperRootPaths[i] + rootBinding.value()[0];
+          }
+        }
+
+        Object subMapper = readDocument(doc, f.getType(), subMapperRootPaths);
+        ReflectionUtils.ensureAccess(f);
+        f.set(bindMapper, subMapper);
+      }
+
+      Set<Field> fieldsWithAnnotation = ReflectionUtils.findFields(bindMapperClass, XPathBinding.class);
+      for (Field f : fieldsWithAnnotation) {
+        XPathBinding fieldAnnotation = f.getAnnotation(XPathBinding.class);
+        ReflectionUtils.ensureAccess(f);
+
+        Object value;
+        Set<String> variables = xPathMapper.getVariables(fieldAnnotation.valueTemplate());
+        Set<String> expressions = new HashSet<>(Arrays.asList(fieldAnnotation.value()));
+        f.set(bindMapper, xPathMapper.evaluateXPathBinding(f.getType(), fieldAnnotation, variables, expressions));
+      }
+
+      return bindMapper;
+    } catch (Throwable t) {       // No more details available!
+      t.printStackTrace();
+      throw new RuntimeException("Cannot instantiate mapper for class " + bindMapperClass.getName() + ": " + t, t);
+    }
+  }
+
+  private static String determineDefaultRootNamespace(XPathRoot rootAnnotation) {
+    String defaultRootNamespace = "";
+    if (rootAnnotation != null && !rootAnnotation.defaultNamespace().isEmpty()) {
+      defaultRootNamespace = rootAnnotation.defaultNamespace();
+    }
+    return defaultRootNamespace;
+  }
+
+  private static String[] determineRootPaths(XPathRoot rootAnnotation, String[] rootPaths) {
+    List<String> rootPathsList = new ArrayList<>();
```

Comment:
Warum hier der ganze Tanz mit dem Konvertieren in Liste und zurück?

```java
if (rootPaths != null) {
  return rootPaths;
} else if (rootAnnotation != null && rootAnnotation.value() != null) {
  return rootAnnotation.value();
} else {
  return null;
}
```

Reply:
Betriebsblindheit... Ist gefixed :)